keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlResourceRepository.kt

```diff
@@ -121,35 +126,50 @@ open class SqlResourceRepository(
   }
 
   // todo: this is not retryable due to overall repository structure: https://github.com/spinnaker/keel/issues/740
-  override fun store(resource: Resource<*>) {
-    val uid = jooq.select(RESOURCE.UID)
+  override fun store(resource: Resource<*>): Resource<*> {
+    val version = jooq.select(
+        coalesce(
+          max(RESOURCE.VERSION),
+          value(0)
+        )
+      )
       .from(RESOURCE)
       .where(RESOURCE.ID.eq(resource.id))
-      .fetchOne(RESOURCE.UID)
-      ?: randomUID().toString()
-
-    val updatePairs = mapOf(
-      RESOURCE.KIND to resource.kind,
-      RESOURCE.ID to resource.id,
-      RESOURCE.SPEC to objectMapper.writeValueAsString(resource.spec),
-      RESOURCE.APPLICATION to resource.application
-    )
-    val insertPairs = updatePairs + (RESOURCE.UID to uid)
-    jooq.insertInto(
-      RESOURCE,
-      *insertPairs.keys.toTypedArray()
-    )
-      .values(*insertPairs.values.toTypedArray())
-      .onDuplicateKeyUpdate()
-      .set(updatePairs)
+      .fetchOneInto(Int::class.java)
+
+    val oldUid = if (version > 0 ) {
+      getResourceUid(resource.id, version)
+    } else {
+      null
+    }
+    val uid = randomUID().toString()
+
+    jooq.insertInto(RESOURCE)
+      .set(RESOURCE.UID, uid)
+      .set(RESOURCE.KIND, resource.kind.toString())
+      .set(RESOURCE.ID, resource.id)
+      .set(RESOURCE.VERSION, version + 1)
+      .set(RESOURCE.SPEC, objectMapper.writeValueAsString(resource.spec))
+      .set(RESOURCE.APPLICATION, resource.application)
       .execute()
+
     jooq.insertInto(RESOURCE_LAST_CHECKED)
       .set(RESOURCE_LAST_CHECKED.RESOURCE_UID, uid)
       .set(RESOURCE_LAST_CHECKED.AT, EPOCH.plusSeconds(1))
-      .onDuplicateKeyUpdate()
-      .set(RESOURCE_LAST_CHECKED.AT, EPOCH.plusSeconds(1))
-      .set(RESOURCE_LAST_CHECKED.IGNORE, false)
       .execute()
+
+    // This is somewhat temporary because eventually we'll define a new environment version
+    if (oldUid != null) {
+      jooq
+        .update(ENVIRONMENT_RESOURCE)
+        .set(ENVIRONMENT_RESOURCE.RESOURCE_UID, uid)
+        .where(ENVIRONMENT_RESOURCE.RESOURCE_UID.eq(oldUid))
+        .execute()
+    }
```

Comment:
If this is an update, re-link the new version of the resource to any environments it belongs to